Index: sqlalchemy_jsonapi/serializer.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sqlalchemy_jsonapi/serializer.py	(revision 3312805c1ffbdd2f3f11f3fed24ecf2bcf277023)
+++ sqlalchemy_jsonapi/serializer.py	(revision 6103e6e3ff3ba00f9a037154244de89763298f9e)
@@ -5,13 +5,15 @@
 MIT License
 """
 
-from collections import MutableMapping
 from enum import Enum
-from inflection import pluralize, dasherize, parameterize, tableize, underscore
+from inflection import dasherize, tableize, underscore
 from sqlalchemy.exc import IntegrityError
+from sqlalchemy.orm import RelationshipProperty
+from sqlalchemy.orm import joinedload
+from sqlalchemy.orm.collections import MappedCollection
 from sqlalchemy.orm.interfaces import MANYTOONE
+from sqlalchemy.sql import operators
 from sqlalchemy.util.langhelpers import iterate_attributes
-from pprint import pprint
 
 from .errors import (BadRequestError, InvalidTypeForEndpointError,
                      MissingTypeError, NotSortableError, PermissionDeniedError,
@@ -303,6 +305,21 @@
         check_permission(resource, relationship.key, permission)
         return relationship
 
+    def _apply_filter(self, collection, model, attr_path, op, value):
+
+        col = None
+
+        for attr in attr_path.split('.'):
+            col = getattr(model, attr)
+            if isinstance(col.prop, RelationshipProperty):
+                model = col.prop.mapper.class_
+                collection = collection.join(col)
+
+        if col:
+            collection = collection.filter(getattr(operators, op)(col, value))
+
+        return collection
+
     def _check_json_data(self, json_data):
         """
         Ensure that the request body is both a hash and has a data key.
@@ -374,13 +391,14 @@
             except PermissionDeniedError:
                 continue
 
-            if relationship.direction == MANYTOONE:
+            if relationship.direction == MANYTOONE or not relationship.uselist:
                 if key in local_fields:
                     to_ret['relationships'][api_key] = {
                         'links': self._lazy_relationship(api_type, instance.id,
                                                          api_key)
                     }
 
+
                 if api_key in include.keys():
                     related = desc(instance)
                     if related is not None:
@@ -410,6 +428,8 @@
                     to_ret['relationships'][api_key]['data'] = []
 
                 related = desc(instance)
+                if isinstance(related, MappedCollection):
+                    related = related.values()
 
                 for item in related:
                     try:
@@ -466,7 +486,7 @@
         field_args = {
             k: v
             for k, v in query.items() if k.startswith('fields[')
-        }
+            }
 
         fields = {}
 
@@ -475,6 +495,24 @@
 
         return fields
 
+    def _parse_filter(self, query):
+        """
+        Parse the querystring args for filters.
+
+        :param query: Dict of query args
+        """
+        filter_args = {
+            k: v
+            for k, v in query.items() if k.startswith('filter[')
+            }
+
+        filters = {}
+        for k, v in filter_args.items():
+            attr, op = k[7:-1].rsplit('.', 1)
+            filters[attr] = op, None if v == 'None' else v
+
+        return filters
+
     def _parse_include(self, include):
         """
         Parse the querystring args or parent includes for includes.
@@ -614,12 +652,24 @@
         model = self._fetch_model(api_key)
         include = self._parse_include(query.get('include', '').split(','))
         fields = self._parse_fields(query)
+        filter = self._parse_filter(query)
         included = {}
         sorts = query.get('sort', '').split(',')
         order_by = []
 
         collection = session.query(model)
 
+        for local in include.keys():
+            if local:
+                loader = joinedload(underscore(local))
+                for remote in include[local]:
+                    loader.joinedload(*[underscore(r) for r in remote.split('.')])
+                collection = collection.options(loader)
+
+        if filter:
+            for attr_path, expr in filter.items():
+                collection = self._apply_filter(collection, model, attr_path, expr[0], expr[1])
+
         for attr in sorts:
             if attr == '':
                 break
@@ -711,6 +761,8 @@
 
         related = get_rel_desc(resource, relationship.key,
                                RelationshipActions.GET)(resource)
+        if isinstance(related, MappedCollection):
+            related = related.values()
 
         if relationship.direction == MANYTOONE:
             try:
@@ -775,7 +827,7 @@
 
         return response
 
-    def patch_relationship(self, session, json_data, api_type, obj_id,
+    def patch_relationship(self, session, query, json_data, api_type, obj_id,
                            rel_key):
         """
         Replacement of relationship values.
@@ -854,10 +906,10 @@
         except KeyError:
             raise ValidationError('Incompatible Type')
 
-        return self.get_relationship(session, {}, model.__jsonapi_type__,
+        return self.get_relationship(session, query, model.__jsonapi_type__,
                                      resource.id, rel_key)
 
-    def patch_resource(self, session, json_data, api_type, obj_id):
+    def patch_resource(self, session, query, json_data, api_type, obj_id):
         """
         Replacement of resource values.
 
@@ -905,7 +957,7 @@
                     continue
 
                 self.patch_relationship(
-                    session, json_data['data']['relationships'][api_key],
+                    session, query, json_data['data']['relationships'][api_key],
                     model.__jsonapi_type__, resource.id, api_key)
 
             data_keys = set(map((lambda x: resource.__jsonapi_map_to_py__.get(x, None)), json_data['data']['attributes'].keys()))
@@ -931,9 +983,9 @@
             session.rollback()
             raise ValidationError('Incompatible data type')
         return self.get_resource(
-            session, {}, model.__jsonapi_type__, resource.id)
+            session, query, model.__jsonapi_type__, resource.id)
 
-    def post_collection(self, session, data, api_type):
+    def post_collection(self, session, query, data, api_type):
         """
         Create a new Resource.
 
@@ -1067,7 +1119,7 @@
             raise ValidationError('Incompatible data type')
         session.refresh(resource)
         response = self.get_resource(
-            session, {}, model.__jsonapi_type__, resource.id)
+            session, query, model.__jsonapi_type__, resource.id)
         response.status_code = 201
         return response

